src/backtracking/hamiltonian_cycle.rs

```diff
@@ -0,0 +1,248 @@
+        let mut path = vec![usize::MAX; self.adjacency_matrix.len()];
+        // Start at the specified vertex.
+        path[0] = start_vertex;
+
+        if self.hamiltonian_cycle_util(&mut path, 1) {
+            // Complete the cycle by returning to the starting vertex.
+            path.push(start_vertex);
+            Ok(Some(path))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// Attempts to find a Hamiltonian cycle in a graph represented by an adjacency matrix, starting from a specified vertex.
+pub fn find_hamiltonian_cycle(
+    adjacency_matrix: Vec<Vec<bool>>,
+    start_vertex: usize,
+) -> Result<Option<Vec<usize>>, AdjMatError> {
+    let graph = Graph::new(adjacency_matrix)?;
+    graph.find_hamiltonian_cycle(start_vertex)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! hamiltonian_cycle_tests {
+        ($($name:ident: $test_case:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (adjacency_matrix, start_vertex, expected) = $test_case;
+                    let result = find_hamiltonian_cycle(adjacency_matrix, start_vertex);
+                    assert_eq!(result, expected);
+                }
+            )*
+        };
+    }
+
+    hamiltonian_cycle_tests! {
+        test_complete_graph: (
+            vec![
+                vec![false, true, true, true],
+                vec![true, false, true, true],
+                vec![true, true, false, true],
+                vec![true, true, true, false],
+            ],
+            0,
+            Ok(Some(vec![0, 1, 2, 3, 0]))
+        ),
+        test_cycle_graph: (
+            vec![
+                vec![false, true, false, false, true],
+                vec![true, false, true, false, false],
+                vec![false, true, false, true, false],
+                vec![false, false, true, false, true],
+                vec![true, false, false, true, false],
+            ],
+            2,
+            Ok(Some(vec![2, 1, 0, 4, 3, 2]))
+        ),
+        test_no_cycle_graph: (
+            vec![
+                vec![false, true, false],
+                vec![true, false, false],
+                vec![false, false, false],
+            ],
+            0,
+            Ok(None::<Vec<usize>>)
+        ),
+        test_triangle_graph: (
+            vec![
+                vec![false, true, true],
+                vec![true, false, true],
+                vec![true, true, false],
+            ],
+            1,
+            Ok(Some(vec![1, 0, 2, 1]))
+        ),
+        test_tree_graph: (
+            vec![
+                vec![false, true, false, true, false],
+                vec![true, false, true, true, false],
+                vec![false, true, false, false, false],
+                vec![true, true, false, false, true],
+                vec![false, false, false, true, false],
+            ],
+            0,
+            Ok(None::<Vec<usize>>)
+        ),
```

Comment:
Does this algorithm work for directed graphs? If I am not mistaking, all of the provided adjacency matrices are symmetric, so the represented graph is undirected.

Reply:
Currently, the implementation only supports undirected graphs.

Reply:
Test for directed graph passed. You can see these tests in the implementation.